pkg/currency/providers.go

```diff
@@ -0,0 +1,37 @@
+package currency
+
+import (
+	"context"
+	"time"
+
+	"github.com/justtrackio/gosoline/pkg/http"
+	"github.com/justtrackio/gosoline/pkg/log"
+)
+
+type ProviderSettings struct {
+	ApiKey string `cfg:"api_key"`
+	// Providers with lower values will be used first. The order of the providers with the same priority is non-deterministic.
+	Priority uint `cfg:"priority" validation:"required"`
+}
+
+//go:generate go run github.com/vektra/mockery/v2 --name Provider
+type Provider interface {
+	Name() string
+	GetPriority() int
+	FetchLatestRates(ctx context.Context) ([]Rate, error)
+	FetchHistoricalRates(ctx context.Context, dates []time.Time) (map[time.Time][]Rate, error)
+}
+
+var ProviderRegistry = map[string]ProviderOptions{
+	ECBProviderName:        ECBProviderOption(),
```

Comment:
I would not export the functions creating these providers, it is not required. I would also remove the call and directly pass the function (then they can also be slightly simplified) and just call them `newWhateverProvider`